domain/keymanager/state/state.go

```diff
@@ -33,196 +35,415 @@ func NewState(factory database.TxnRunnerFactory) *State {
+FROM v_model
+WHERE uuid = $modelUUIDValue.model_uuid
+`, modelUUIDVal)
+	if err != nil {
+		return errors.Errorf(
+			"creating model exists statement when checking if model %q exists: %w",
+			modelUUID, err,
+		)
 	}
 
-	db, err := s.DB()
+	err = tx.Query(ctx, modelExistsStmt, modelUUIDVal).Get(&modelUUIDVal)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return errors.Errorf(
+			"checking model %q exists, model not found", modelUUID,
+		).Add(modelerrors.NotFound)
+	} else if err != nil {
+		return errors.Errorf(
+			"checking model %q exists: %w", modelUUID, err,
+		)
+	}
+
+	return nil
+}
+
+// checkUserExists is responsible for checking that a given user exists with the
+// controller. If the user exists then no error is returned otherwise an error
+// satisfying [accesserrors.UserNotFound] is returned.
+func (s *State) checkUserExists(
+	ctx context.Context,
+	userUUID user.UUID,
+	tx *sqlair.TX,
+) error {
+	userUUIDVal := userUUIDValue{UUID: userUUID.String()}
+
+	userExistsStmt, err := s.Prepare(`
+SELECT (uuid) AS (&userUUIDValue.user_uuid)
+FROM v_user_auth
+WHERE uuid = $userUUIDValue.user_uuid
+AND removed = false
+`, userUUIDVal)
 	if err != nil {
-		return errors.Trace(err)
+		return errors.Errorf(
+			"creating user exists statement when checking if user %q exists: %w",
+			userUUID, err,
+		)
 	}
 
-	insertStatement, err := s.Prepare(`
+	err = tx.Query(ctx, userExistsStmt, userUUIDVal).Get(&userUUIDVal)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return errors.Errorf(
+			"checking user %q exists, user not found",
+			userUUID,
+		).Add(accesserrors.UserNotFound)
+	} else if err != nil {
+		return errors.Errorf(
+			"checking user %q exists: %w",
+			userUUID, err,
+		)
+	}
+
+	return nil
+}
+
+// ensureUserPublicKey ensures that a public key exists for a user returning the
+// unique id to represent that key within the database.
+func (s *State) ensureUserPublicKey(
+	ctx context.Context,
+	key userPublicKeyInsert,
+	tx *sqlair.TX,
+) (int64, error) {
+	insertStmt, err := s.Prepare(`
 INSERT INTO user_public_ssh_key (comment,
                                  fingerprint,
 						         public_key,
-                                 user_id,
+                                 user_uuid,
                                  fingerprint_hash_algorithm_id)
 SELECT $userPublicKeyInsert.comment,
        $userPublicKeyInsert.fingerprint,
        $userPublicKeyInsert.public_key,
-       $userPublicKeyInsert.user_id,
+       $userPublicKeyInsert.user_uuid,
        s.id	
 FROM ssh_fingerprint_hash_algorithm s
 WHERE s.algorithm = $userPublicKeyInsert.algorithm
 `, userPublicKeyInsert{})
 
 	if err != nil {
-		return fmt.Errorf(
-			"preparing insert statement for adding user %q public keys: %w",
-			userId,
+		return 0, errors.Errorf(
+			"preparing insert statement for ensuring user public key: %w",
 			err,
 		)
 	}
 
+	selectExistingIdStmt, err := s.Prepare(`
+SELECT (id) AS (&userPublicKeyId.*)
```

Comment:
`SELECT &userPubicKeyId.id`

(and same for anywhere else where AS is not needed)

Reply:
Done & Done